File: nxflix-agents-python/src/nxflix_agents/routers/kanji.py
"""Kanji API routes."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from nxflix_agents.models import KanjiItem, KanjiGenerateRequest
from nxflix_agents.services import KanjiService
from nxflix_agents.providers.llm import LLMProvider

router = APIRouter(prefix="/api/kanji", tags=["kanji"])

# Singleton kanji service
_kanji_service = KanjiService()
_llm = LLMProvider()


class KanjiListResponse(BaseModel):
    """Response for listing kanji."""

    kanji: list[KanjiItem]
    count: int


class KanjiSingleResponse(BaseModel):
    """Response for a single kanji."""

    kanji: KanjiItem


class KanjiGeneratedResponse(BaseModel):
    """Response schema for generated kanji."""

    kanji: list[KanjiItem]


@router.get("/", response_model=KanjiListResponse)
async def list_kanji():
    """List all kanji items."""
    kanji = _kanji_service.get_all_kanji_items()
    return KanjiListResponse(kanji=kanji, count=len(kanji))


@router.get("/search", response_model=KanjiListResponse)
async def search_kanji(query: str):
    """Search kanji by character, meaning, or readings."""
    kanji = _kanji_service.search_kanji(query)
    return KanjiListResponse(kanji=kanji, count=len(kanji))


@router.get("/by-character/{char}", response_model=KanjiSingleResponse)
async def get_kanji_by_character(char: str):
    """Get kanji by its character."""
    kanji = _kanji_service.get_kanji_by_character(char)
    if not kanji:
        raise HTTPException(status_code=404, detail="Kanji not found")
    return KanjiSingleResponse(kanji=kanji)


@router.get("/{kanji_id}", response_model=KanjiSingleResponse)
async def get_kanji(kanji_id: str):
    """Get a single kanji by ID."""
    kanji = _kanji_service.get_kanji_item(kanji_id)
    if not kanji:
        raise HTTPException(status_code=404, detail="Kanji not found")
    return KanjiSingleResponse(kanji=kanji)


@router.post("/generate", response_model=KanjiListResponse)
async def generate_kanji(request: KanjiGenerateRequest):
    """AI-generate a set of kanji items."""
    prompt = _build_kanji_generation_prompt(request)

    try:
        generated = await _llm.complete_json(
            messages=[{"role": "user", "content": prompt}],
            response_model=KanjiGeneratedResponse,
        )

        # Add generated kanji to service
        for kanji in generated.kanji:
            _kanji_service.add_kanji_item(kanji)

        return KanjiListResponse(kanji=generated.kanji, count=len(generated.kanji))

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


def _build_kanji_generation_prompt(request: KanjiGenerateRequest) -> str:
    """Build the prompt for kanji generation."""
    parts: list[str] = [
        "Generate JLPT N1 kanji data in JSON format.",
        "",
    ]

    if request.characters and len(request.characters) > 0:
        parts.append(f"Generate data for these specific kanji: {', '.join(request.characters)}")
    elif request.topic:
        parts.append(f"Generate {request.count} kanji related to the topic: {request.topic}")
    else:
        parts.append(f"Generate {request.count} random JLPT N1 kanji.")

    parts.extend([
        "",
        "For each kanji, provide:",
        "- id: unique identifier (format: kanji-XXX)",
        "- character: the single kanji character",
        "- stroke_count: number of strokes",
        "- onyomi: array of on readings in katakana",
        "- kunyomi: array of kun readings in hiragana",
        "- meanings: array of English meanings",
        "- radicals: array of radical components",
    ])

    if request.include_compounds:
        parts.append("- compound_words: array of 2-3 compound words with word, reading, and meaning")

    parts.extend([
        "- mnemonics: a helpful memory aid (optional)",
        "",
        'Return a JSON object with a "kanji" array containing the kanji objects.',
    ])

    return "\n".join(parts)


# Export the service for state management
def get_kanji_service() -> KanjiService:
    """Get the kanji service singleton."""
    return _kanji_service

File: nxflix-agents-ts/src/routers/kanji.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { kanjiRepository } from '../db/repositories/index.js';
import { KanjiItem, KanjiGenerateRequest, type KanjiItem as KanjiItemType } from '../models/kanji.js';
import { LLMProvider } from '../providers/llm.js';
import type { Kanji as DbKanji } from '../db/schema.js';

const llm = new LLMProvider();
const kanjiRouter = Router();

// Convert database record to model type
function dbToModel(db: DbKanji): KanjiItemType {
  return {
    id: db.id,
    character: db.character,
    strokeCount: db.strokeCount,
    onyomi: db.onyomi,
    kunyomi: db.kunyomi,
    meanings: db.meanings,
    radicals: db.radicals || [],
    compoundWords: db.compoundWords || [],
    mnemonics: db.mnemonics || undefined,
    level: db.level,
    contentType: 'kanji',
  };
}

// Request schemas
const KanjiSearchSchema = z.object({
  query: z.string().min(1),
});

const KanjiGenerateSchema = KanjiGenerateRequest;

// Response types
interface KanjiListResponse {
  kanji: KanjiItem[];
  count: number;
}

interface KanjiSingleResponse {
  kanji: KanjiItem;
}

// GET /api/kanji - List all kanji items
kanjiRouter.get('/', async (_req: Request, res: Response) => {
  try {
    const dbResults = await kanjiRepository.findAll();
    const kanji = dbResults.map(dbToModel);
    res.json({ kanji, count: kanji.length } as KanjiListResponse);
  } catch (error) {
    console.error('Error fetching kanji:', error);
    res.status(500).json({ error: String(error) });
  }
});

// GET /api/kanji/search - Search kanji
kanjiRouter.get('/search', async (req: Request, res: Response) => {
  try {
    const { query } = KanjiSearchSchema.parse(req.query);
    const dbResults = await kanjiRepository.search(query);
    const kanji = dbResults.map(dbToModel);
    res.json({ kanji, count: kanji.length } as KanjiListResponse);
  } catch (error) {
    console.error('Error searching kanji:', error);
    res.status(400).json({ error: String(error) });
  }
});

// GET /api/kanji/by-character/:char - Get kanji by character
kanjiRouter.get('/by-character/:char', async (req: Request<{ char: string }>, res: Response) => {
  try {
    const dbResult = await kanjiRepository.findByCharacter(req.params.char);
    if (!dbResult) {
      res.status(404).json({ error: 'Kanji not found' });
      return;
    }
    res.json({ kanji: dbToModel(dbResult) } as KanjiSingleResponse);
  } catch (error) {
    console.error('Error fetching kanji:', error);
    res.status(500).json({ error: String(error) });
  }
});

// GET /api/kanji/:id - Get single kanji by ID
kanjiRouter.get('/:id', async (req: Request<{ id: string }>, res: Response) => {
  try {
    const dbResult = await kanjiRepository.findById(req.params.id);
    if (!dbResult) {
      res.status(404).json({ error: 'Kanji not found' });
      return;
    }
    res.json({ kanji: dbToModel(dbResult) } as KanjiSingleResponse);
  } catch (error) {
    console.error('Error fetching kanji:', error);
    res.status(500).json({ error: String(error) });
  }
});

// POST /api/kanji/generate - AI-generate kanji set (does NOT save)
kanjiRouter.post('/generate', async (req: Request, res: Response) => {
  try {
    const request = KanjiGenerateSchema.parse(req.body);

    const prompt = buildKanjiGenerationPrompt(request);

    const generatedKanji = await llm.completeJson(
      [{ role: 'user', content: prompt }],
      z.object({
        kanji: z.array(KanjiItem),
      })
    );

    // Return generated content without saving
    res.json({ kanji: generatedKanji.kanji, count: generatedKanji.kanji.length } as KanjiListResponse);
  } catch (error) {
    console.error('Error generating kanji:', error);
    res.status(500).json({ error: String(error) });
  }
});

// POST /api/kanji/save - Save kanji to database
kanjiRouter.post('/save', async (req: Request, res: Response) => {
  try {
    const { kanji, isPublic, userId } = req.body as {
      kanji: KanjiItemType[];
      isPublic?: boolean;
      userId?: string;
    };
    if (!kanji || !Array.isArray(kanji)) {
      res.status(400).json({ error: 'kanji array is required' });
      return;
    }

    const savedKanji: KanjiItemType[] = [];
    for (const item of kanji) {
      const saved = await kanjiRepository.upsert({
        id: item.id,
        character: item.character,
        strokeCount: item.strokeCount,
        onyomi: item.onyomi,
        kunyomi: item.kunyomi,
        meanings: item.meanings,
        radicals: item.radicals || [],
        compoundWords: item.compoundWords || [],
        mnemonics: item.mnemonics,
        level: item.level || 'N1',
        isPublic: isPublic ?? false,
        createdBy: userId,
      });
      savedKanji.push(dbToModel(saved));
    }

    res.json({ kanji: savedKanji, count: savedKanji.length, saved: true } as KanjiListResponse & { saved: boolean });
  } catch (error) {
    console.error('Error saving kanji:', error);
    res.status(500).json({ error: String(error) });
  }
});

function buildKanjiGenerationPrompt(request: z.infer<typeof KanjiGenerateSchema>): string {
  const parts: string[] = [
    'Generate JLPT N1 kanji data in JSON format.',
    '',
  ];

  if (request.characters && request.characters.length > 0) {
    parts.push(`Generate data for these specific kanji: ${request.characters.join(', ')}`);
  } else if (request.topic) {
    parts.push(`Generate ${request.count} kanji related to the topic: ${request.topic}`);
  } else {
    parts.push(`Generate ${request.count} random JLPT N1 kanji.`);
  }

  parts.push('');
  parts.push('For each kanji, provide:');
  parts.push('- id: unique identifier (format: kanji-XXX)');
  parts.push('- character: the single kanji character');
  parts.push('- strokeCount: number of strokes');
  parts.push('- onyomi: array of on readings in katakana');
  parts.push('- kunyomi: array of kun readings in hiragana');
  parts.push('- meanings: array of English meanings');
  parts.push('- radicals: array of radical components');

  if (request.includeCompounds) {
    parts.push('- compoundWords: array of 2-3 compound words with word, reading, and meaning');
  }

  parts.push('- mnemonics: a helpful memory aid (optional)');
  parts.push('');
  parts.push('Return a JSON object with a "kanji" array containing the kanji objects.');

  return parts.join('\n');
}

export { kanjiRouter };
